swap_server_client.go

```diff
@@ -275,8 +301,151 @@ func (s *grpcSwapServerClient) NewLoopInSwap(ctx context.Context,
+			ProtocolVersion: protocolVersion,
+			SwapHash:        hash[:],
+		},
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	receive := func() (*ServerUpdate, error) {
+		response, err := resp.Recv()
+		if err != nil {
+			return nil, err
+		}
+
+		return &ServerUpdate{
+			State:     response.State,
+			Timestamp: time.Unix(0, response.TimestampNs),
+		}, err
+	}
+
+	updateChan, errChan := s.makeServerUpdate(ctx, receive)
+	return updateChan, errChan, nil
+}
+
+// SubscribeLoopOutUpdates subscribes to loop out server state and pipes updates
+// into the channel provided.
+func (s *grpcSwapServerClient) SubscribeLoopOutUpdates(ctx context.Context,
+	hash lntypes.Hash) (<-chan *ServerUpdate, <-chan error, error) {
+
+	resp, err := s.server.SubscribeLoopOutUpdates(
+		ctx, &looprpc.SubscribeUpdatesRequest{
+			ProtocolVersion: protocolVersion,
+			SwapHash:        hash[:],
+		},
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	receive := func() (*ServerUpdate, error) {
+		response, err := resp.Recv()
+		if err != nil {
+			return nil, err
+		}
+
+		return &ServerUpdate{
+			State:     response.State,
+			Timestamp: time.Unix(0, response.TimestampNs),
+		}, err
+	}
+
+	updateChan, errChan := s.makeServerUpdate(ctx, receive)
+	return updateChan, errChan, nil
+}
+
+// makeServerUpdate takes a stream receive function and a channel that it
+// should pipe updates into. It sends events into the updates channel until
+// the client cancels, server client shuts down or the subscription is cancelled
+// server side.
+func (s *grpcSwapServerClient) makeServerUpdate(ctx context.Context,
+	receive func() (*ServerUpdate, error)) (<-chan *ServerUpdate,
+	<-chan error) {
+
+	// We will return exactly one error from this function so we buffer
+	// our error channel so that the function exit is not dependent on
+	// the error being read.
+	errChan := make(chan error, 1)
+	updateChan := make(chan *ServerUpdate)
+
+	// Create a goroutine that will pipe updates in to our updates channel.
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+
+		for {
+			// Try to receive from our stream. If there are no items
+			// to consume, this call will block. If our stream is
+			// cancelled by the server we will receive an error.
+			response, err := receive()
+			switch err {
+			// If we get a nil error, we fallthrough to deliver
+			// the update we have just received.
+			case nil:
+
+			// If we get an EOF error, the server is finished
+			// sending us updates, so we return with a non-nil
+			// error. This is possible for swaps where the server
+			// marks the swap as a success before we do.
+			case io.EOF:
+				log.Trace("server subscription finished")
+
+				errChan <- nil
+				return
+
+			// If we receive a non-nil error, we exit.
+			default:
+				// If we get a transport is closing error, we
+				// return a nil error because the server is
+				// restarting.
+				if isErrConClosing(err) {
+					log.Trace("server subscription exit")
```

Comment:
Same comment about trace. This is probably important to show to the user, so that they know they won't be kept up to date anymore.

Reply:
Maybe it isn't necessary (and also tricky) to do string match and the error can be passed back as is?

Reply:
I really don't like exposing rpc errors to a user, they're very unfriendly. Would rather return a nicely formatted shutting down error.

Reply:
Ok, yes remapping the error is also a possibility. Although we don't know if it is really the server shutting down or the user's internet having a hick up.

Reply:
What do you think about this? The current error message seems to narrow for the range of events that can trigger this

Reply:
Can make it more generic? "subscription failed, no further updates will be sent"

Reply:
Sounds good to me. At least a lot less specific than 'server restart'